File: Algorithms/073_SetMatrixZeroes.java
public class Solution {
    public void setZeroes(int[][] matrix) {
        int M = matrix.length;
        int N = matrix[0].length;
        boolean setFirstRowZeroes = false;
        boolean setFirstColZeroes = false;
        for (int i=0; i < M && !setFirstColZeroes; i++) {
            if (matrix[i][0] == 0) {
                setFirstColZeroes = true;
            }
        }
        for (int j=0; j < N && !setFirstRowZeroes; j++) {
            if (matrix[0][j] == 0) {
                setFirstRowZeroes = true;
            }
        }
        for (int i=1; i < M; i++) {
            for (int j=1; j < N; j++) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }
        for (int i=1; i < M; i++) {
            if (matrix[i][0] == 0) {
                for (int j=1; j < N; j++) {
                    matrix[i][j] = 0;
                }
            }
        }
        for (int j=1; j < N; j++) {
            if (matrix[0][j] == 0) {
                for (int i=1; i < M; i++) {
                    matrix[i][j] = 0;
                }
            }
        }
        if (setFirstRowZeroes) {
            for (int j=0; j < N; j++) {
                matrix[0][j] = 0;
            }
        }
        if (setFirstColZeroes) {
            for (int i=0; i < M; i++) {
                matrix[i][0] = 0;
            }
        }
    }
}

File: Algorithms/073_SetMatrixZeroes.py
class Solution:
    # @param matrix, a list of lists of integers
    # RETURN NOTHING, MODIFY matrix IN PLACE.
    def setZeroes(self, matrix):
        M = len(matrix)
        N = len(matrix[0])
        setFirstRowZeroes = False
        setFirstColZeroes = False
        for i in range(M):
            if matrix[i][0] == 0:
                setFirstColZeroes = True
                break
        for j in range(N):
            if matrix[0][j] == 0:
                setFirstRowZeroes = True
                break
        for i in range(1, M):
            for j in range(1, N):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0
        for i in range(1, M):
            if matrix[i][0] == 0:
                for j in range(1, N):
                    matrix[i][j] = 0
        for j in range(1, N):
            if matrix[0][j] == 0:
                for i in range(1, M):
                    matrix[i][j] = 0
        if setFirstRowZeroes:
            for j in range(N):
                matrix[0][j] = 0
        if setFirstColZeroes:
            for i in range(M):
                matrix[i][0] = 0
        return matrix
